File: scitt-golang/pkg/database/log_state_test.go
package database_test

import (
	"path/filepath"
	"testing"

	"github.com/tradeverifyd/transparency-service/scitt-golang/pkg/database"
)

func TestGetCurrentTreeSize(t *testing.T) {
	t.Run("returns initial tree size of 0", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		size, err := database.GetCurrentTreeSize(db)
		if err != nil {
			t.Fatalf("failed to get tree size: %v", err)
		}

		if size != 0 {
			t.Errorf("expected tree size 0, got %d", size)
		}
	})

	t.Run("returns updated tree size", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		// Update tree size
		err = database.UpdateTreeSize(db, 100)
		if err != nil {
			t.Fatalf("failed to update tree size: %v", err)
		}

		// Get tree size
		size, err := database.GetCurrentTreeSize(db)
		if err != nil {
			t.Fatalf("failed to get tree size: %v", err)
		}

		if size != 100 {
			t.Errorf("expected tree size 100, got %d", size)
		}
	})
}

func TestUpdateTreeSize(t *testing.T) {
	t.Run("updates tree size", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		err = database.UpdateTreeSize(db, 42)
		if err != nil {
			t.Fatalf("failed to update tree size: %v", err)
		}

		size, err := database.GetCurrentTreeSize(db)
		if err != nil {
			t.Fatalf("failed to get tree size: %v", err)
		}

		if size != 42 {
			t.Errorf("expected tree size 42, got %d", size)
		}
	})

	t.Run("updates tree size multiple times", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		sizes := []int64{10, 20, 30, 40}

		for _, expectedSize := range sizes {
			err = database.UpdateTreeSize(db, expectedSize)
			if err != nil {
				t.Fatalf("failed to update tree size: %v", err)
			}

			size, err := database.GetCurrentTreeSize(db)
			if err != nil {
				t.Fatalf("failed to get tree size: %v", err)
			}

			if size != expectedSize {
				t.Errorf("expected tree size %d, got %d", expectedSize, size)
			}
		}
	})
}

func TestRecordTreeState(t *testing.T) {
	t.Run("records tree state", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		state := database.TreeState{
			TreeSize:             100,
			RootHash:             "abc123",
			CheckpointStorageKey: "checkpoints/100",
			CheckpointSignedNote: "signed note content",
		}

		err = database.RecordTreeState(db, state)
		if err != nil {
			t.Fatalf("failed to record tree state: %v", err)
		}

		// Verify state was recorded
		retrieved, err := database.GetTreeState(db, 100)
		if err != nil {
			t.Fatalf("failed to get tree state: %v", err)
		}

		if retrieved == nil {
			t.Fatal("tree state not found")
		}

		if retrieved.TreeSize != state.TreeSize {
			t.Errorf("expected tree size %d, got %d", state.TreeSize, retrieved.TreeSize)
		}
		if retrieved.RootHash != state.RootHash {
			t.Errorf("expected root hash %s, got %s", state.RootHash, retrieved.RootHash)
		}
	})

	t.Run("records multiple tree states", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		states := []database.TreeState{
			{TreeSize: 100, RootHash: "hash1", CheckpointStorageKey: "cp1", CheckpointSignedNote: "note1"},
			{TreeSize: 200, RootHash: "hash2", CheckpointStorageKey: "cp2", CheckpointSignedNote: "note2"},
			{TreeSize: 300, RootHash: "hash3", CheckpointStorageKey: "cp3", CheckpointSignedNote: "note3"},
		}

		for _, state := range states {
			err = database.RecordTreeState(db, state)
			if err != nil {
				t.Fatalf("failed to record tree state: %v", err)
			}
		}

		// Verify all states were recorded
		for _, state := range states {
			retrieved, err := database.GetTreeState(db, state.TreeSize)
			if err != nil {
				t.Fatalf("failed to get tree state: %v", err)
			}

			if retrieved == nil {
				t.Errorf("tree state %d not found", state.TreeSize)
				continue
			}

			if retrieved.RootHash != state.RootHash {
				t.Errorf("tree size %d: expected root hash %s, got %s", state.TreeSize, state.RootHash, retrieved.RootHash)
			}
		}
	})
}

func TestGetTreeState(t *testing.T) {
	t.Run("returns nil for non-existent tree state", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		state, err := database.GetTreeState(db, 999)
		if err != nil {
			t.Fatalf("failed to get tree state: %v", err)
		}

		if state != nil {
			t.Error("expected nil for non-existent tree state")
		}
	})

	t.Run("retrieves existing tree state", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		original := database.TreeState{
			TreeSize:             500,
			RootHash:             "xyz789",
			CheckpointStorageKey: "checkpoints/500",
			CheckpointSignedNote: "checkpoint note",
		}

		err = database.RecordTreeState(db, original)
		if err != nil {
			t.Fatalf("failed to record tree state: %v", err)
		}

		retrieved, err := database.GetTreeState(db, 500)
		if err != nil {
			t.Fatalf("failed to get tree state: %v", err)
		}

		if retrieved == nil {
			t.Fatal("tree state not found")
		}

		if retrieved.TreeSize != original.TreeSize {
			t.Errorf("expected tree size %d, got %d", original.TreeSize, retrieved.TreeSize)
		}
		if retrieved.RootHash != original.RootHash {
			t.Errorf("expected root hash %s, got %s", original.RootHash, retrieved.RootHash)
		}
		if retrieved.CheckpointStorageKey != original.CheckpointStorageKey {
			t.Errorf("expected checkpoint key %s, got %s", original.CheckpointStorageKey, retrieved.CheckpointStorageKey)
		}
		if retrieved.CheckpointSignedNote != original.CheckpointSignedNote {
			t.Errorf("expected signed note %s, got %s", original.CheckpointSignedNote, retrieved.CheckpointSignedNote)
		}
	})
}

func TestGetTreeStateHistory(t *testing.T) {
	t.Run("returns empty history for new database", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		history, err := database.GetTreeStateHistory(db, 0)
		if err != nil {
			t.Fatalf("failed to get tree state history: %v", err)
		}

		if len(history) != 0 {
			t.Errorf("expected empty history, got %d states", len(history))
		}
	})

	t.Run("returns history in descending order", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		// Record states
		states := []database.TreeState{
			{TreeSize: 100, RootHash: "hash1", CheckpointStorageKey: "cp1", CheckpointSignedNote: "note1"},
			{TreeSize: 200, RootHash: "hash2", CheckpointStorageKey: "cp2", CheckpointSignedNote: "note2"},
			{TreeSize: 300, RootHash: "hash3", CheckpointStorageKey: "cp3", CheckpointSignedNote: "note3"},
		}

		for _, state := range states {
			err = database.RecordTreeState(db, state)
			if err != nil {
				t.Fatalf("failed to record tree state: %v", err)
			}
		}

		// Get history
		history, err := database.GetTreeStateHistory(db, 0)
		if err != nil {
			t.Fatalf("failed to get tree state history: %v", err)
		}

		if len(history) != 3 {
			t.Fatalf("expected 3 states, got %d", len(history))
		}

		// Verify descending order
		if history[0].TreeSize != 300 {
			t.Errorf("expected first tree size 300, got %d", history[0].TreeSize)
		}
		if history[1].TreeSize != 200 {
			t.Errorf("expected second tree size 200, got %d", history[1].TreeSize)
		}
		if history[2].TreeSize != 100 {
			t.Errorf("expected third tree size 100, got %d", history[2].TreeSize)
		}
	})

	t.Run("respects limit parameter", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		// Record 5 states
		for i := 1; i <= 5; i++ {
			state := database.TreeState{
				TreeSize:             int64(i * 100),
				RootHash:             "hash",
				CheckpointStorageKey: "cp",
				CheckpointSignedNote: "note",
			}
			err = database.RecordTreeState(db, state)
			if err != nil {
				t.Fatalf("failed to record tree state: %v", err)
			}
		}

		// Get limited history
		history, err := database.GetTreeStateHistory(db, 2)
		if err != nil {
			t.Fatalf("failed to get tree state history: %v", err)
		}

		if len(history) != 2 {
			t.Errorf("expected 2 states with limit, got %d", len(history))
		}

		// Should get the 2 most recent
		if history[0].TreeSize != 500 {
			t.Errorf("expected first tree size 500, got %d", history[0].TreeSize)
		}
		if history[1].TreeSize != 400 {
			t.Errorf("expected second tree size 400, got %d", history[1].TreeSize)
		}
	})
}

func TestGetLatestCheckpoint(t *testing.T) {
	t.Run("returns nil when no checkpoints exist", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		checkpoint, err := database.GetLatestCheckpoint(db)
		if err != nil {
			t.Fatalf("failed to get latest checkpoint: %v", err)
		}

		if checkpoint != nil {
			t.Error("expected nil checkpoint for empty database")
		}
	})

	t.Run("returns most recent checkpoint", func(t *testing.T) {
		tmpDir := t.TempDir()
		dbPath := filepath.Join(tmpDir, "test.db")

		db, err := database.OpenDatabase(database.DatabaseOptions{
			Path:      dbPath,
			EnableWAL: false,
		})
		if err != nil {
			t.Fatalf("failed to open database: %v", err)
		}
		defer database.CloseDatabase(db)

		// Record checkpoints
		states := []database.TreeState{
			{TreeSize: 100, RootHash: "hash1", CheckpointStorageKey: "cp1", CheckpointSignedNote: "note1"},
			{TreeSize: 200, RootHash: "hash2", CheckpointStorageKey: "cp2", CheckpointSignedNote: "note2"},
			{TreeSize: 300, RootHash: "hash3", CheckpointStorageKey: "cp3", CheckpointSignedNote: "note3"},
		}

		for _, state := range states {
			err = database.RecordTreeState(db, state)
			if err != nil {
				t.Fatalf("failed to record tree state: %v", err)
			}
		}

		// Get latest
		latest, err := database.GetLatestCheckpoint(db)
		if err != nil {
			t.Fatalf("failed to get latest checkpoint: %v", err)
		}

		if latest == nil {
			t.Fatal("latest checkpoint not found")
		}

		if latest.TreeSize != 300 {
			t.Errorf("expected latest tree size 300, got %d", latest.TreeSize)
		}
		if latest.RootHash != "hash3" {
			t.Errorf("expected latest root hash hash3, got %s", latest.RootHash)
		}
	})
}

File: scitt-typescript/tests/unit/database/log-state.test.ts
/**
 * Unit tests for Merkle tree state management
 *
 * Tests log state operations including:
 * - Getting current tree size
 * - Updating tree size
 * - Recording tree state at specific sizes
 * - Getting tree state history
 * - Managing checkpoint references
 */

import { describe, test, expect, beforeEach } from "bun:test";
import { Database } from "bun:sqlite";
import * as fs from "fs";
import {
  getCurrentTreeSize,
  updateTreeSize,
  recordTreeState,
  getTreeState,
  getTreeStateHistory,
  getLatestCheckpoint,
} from "../../../src/lib/database/log-state.ts";

const TEST_DB_PATH = "tests/.test-log-state.db";

describe("Log State Management", () => {
  let db: Database;

  beforeEach(() => {
    // Clean up any existing test database
    if (fs.existsSync(TEST_DB_PATH)) {
      fs.unlinkSync(TEST_DB_PATH);
    }

    // Create fresh database with schema
    db = new Database(TEST_DB_PATH);

    // Create tables
    db.exec(`
      CREATE TABLE current_tree_size (
        id INTEGER PRIMARY KEY CHECK (id = 1),
        tree_size INTEGER NOT NULL DEFAULT 0,
        last_updated TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      );

      INSERT INTO current_tree_size (id, tree_size) VALUES (1, 0);

      CREATE TABLE tree_state (
        tree_size INTEGER PRIMARY KEY,
        root_hash TEXT NOT NULL,
        checkpoint_storage_key TEXT NOT NULL,
        checkpoint_signed_note TEXT NOT NULL,
        updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      );
    `);
  });

  test("getCurrentTreeSize returns initial size of 0", () => {
    const size = getCurrentTreeSize(db);
    expect(size).toBe(0);
  });

  test("updateTreeSize increments tree size", () => {
    expect(getCurrentTreeSize(db)).toBe(0);

    updateTreeSize(db, 1);
    expect(getCurrentTreeSize(db)).toBe(1);

    updateTreeSize(db, 2);
    expect(getCurrentTreeSize(db)).toBe(2);

    updateTreeSize(db, 10);
    expect(getCurrentTreeSize(db)).toBe(10);
  });

  test("updateTreeSize updates last_updated timestamp", () => {
    // Get initial timestamp
    updateTreeSize(db, 1);
    const before = db.query(
      "SELECT last_updated FROM current_tree_size WHERE id = 1"
    ).get();

    // Wait to ensure timestamp changes (SQLite CURRENT_TIMESTAMP is second precision)
    Bun.sleepSync(1100); // 1.1 seconds to ensure change

    updateTreeSize(db, 5);

    const after = db.query(
      "SELECT last_updated FROM current_tree_size WHERE id = 1"
    ).get();

    expect(after.last_updated).not.toBe(before.last_updated);
  });

  test("recordTreeState stores tree state with checkpoint", () => {
    const state = {
      tree_size: 10,
      root_hash: "abc123def456",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "example.com/log\n10\nabc123def456\n\n— example.com signature",
    };

    recordTreeState(db, state);

    const row = db.query("SELECT * FROM tree_state WHERE tree_size = ?").get(10);
    expect(row).toBeTruthy();
    expect(row.tree_size).toBe(10);
    expect(row.root_hash).toBe("abc123def456");
    expect(row.checkpoint_storage_key).toBe("checkpoint");
    expect(row.checkpoint_signed_note).toContain("example.com/log");
  });

  test("getTreeState retrieves specific tree state", () => {
    recordTreeState(db, {
      tree_size: 5,
      root_hash: "hash-5",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-5",
    });

    recordTreeState(db, {
      tree_size: 10,
      root_hash: "hash-10",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-10",
    });

    const state5 = getTreeState(db, 5);
    expect(state5).toBeTruthy();
    expect(state5.tree_size).toBe(5);
    expect(state5.root_hash).toBe("hash-5");

    const state10 = getTreeState(db, 10);
    expect(state10).toBeTruthy();
    expect(state10.tree_size).toBe(10);
    expect(state10.root_hash).toBe("hash-10");
  });

  test("getTreeState returns null for non-existent size", () => {
    const state = getTreeState(db, 999);
    expect(state).toBe(null);
  });

  test("getTreeStateHistory returns states in descending order", () => {
    recordTreeState(db, {
      tree_size: 1,
      root_hash: "hash-1",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-1",
    });

    recordTreeState(db, {
      tree_size: 5,
      root_hash: "hash-5",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-5",
    });

    recordTreeState(db, {
      tree_size: 10,
      root_hash: "hash-10",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-10",
    });

    const history = getTreeStateHistory(db);

    expect(history).toHaveLength(3);
    expect(history[0].tree_size).toBe(10);
    expect(history[1].tree_size).toBe(5);
    expect(history[2].tree_size).toBe(1);
  });

  test("getTreeStateHistory with limit", () => {
    for (let i = 1; i <= 10; i++) {
      recordTreeState(db, {
        tree_size: i,
        root_hash: `hash-${i}`,
        checkpoint_storage_key: "checkpoint",
        checkpoint_signed_note: `note-${i}`,
      });
    }

    const history = getTreeStateHistory(db, 3);

    expect(history).toHaveLength(3);
    expect(history[0].tree_size).toBe(10);
    expect(history[1].tree_size).toBe(9);
    expect(history[2].tree_size).toBe(8);
  });

  test("getLatestCheckpoint returns most recent tree state", () => {
    recordTreeState(db, {
      tree_size: 1,
      root_hash: "hash-1",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-1",
    });

    recordTreeState(db, {
      tree_size: 5,
      root_hash: "hash-5",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-5",
    });

    recordTreeState(db, {
      tree_size: 10,
      root_hash: "hash-10",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-10",
    });

    const latest = getLatestCheckpoint(db);

    expect(latest).toBeTruthy();
    expect(latest.tree_size).toBe(10);
    expect(latest.root_hash).toBe("hash-10");
    expect(latest.checkpoint_signed_note).toBe("note-10");
  });

  test("getLatestCheckpoint returns null when no checkpoints exist", () => {
    const latest = getLatestCheckpoint(db);
    expect(latest).toBe(null);
  });

  test("tree_size is primary key (unique)", () => {
    recordTreeState(db, {
      tree_size: 10,
      root_hash: "hash-1",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-1",
    });

    // Attempting to record state with same tree_size should throw
    expect(() => {
      recordTreeState(db, {
        tree_size: 10,
        root_hash: "hash-2",
        checkpoint_storage_key: "checkpoint",
        checkpoint_signed_note: "note-2",
      });
    }).toThrow();
  });

  test("updateTreeSize only allows single row (singleton)", () => {
    // The CHECK constraint ensures only id=1 can exist
    expect(() => {
      db.exec("INSERT INTO current_tree_size (id, tree_size) VALUES (2, 100)");
    }).toThrow();
  });

  test("tree state tracks append-only log growth", () => {
    // Simulate log growing
    updateTreeSize(db, 1);
    recordTreeState(db, {
      tree_size: 1,
      root_hash: "root-1",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "checkpoint-1",
    });

    updateTreeSize(db, 5);
    recordTreeState(db, {
      tree_size: 5,
      root_hash: "root-5",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "checkpoint-5",
    });

    updateTreeSize(db, 10);
    recordTreeState(db, {
      tree_size: 10,
      root_hash: "root-10",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "checkpoint-10",
    });

    // Verify current size
    expect(getCurrentTreeSize(db)).toBe(10);

    // Verify we can retrieve any historical state
    expect(getTreeState(db, 1).root_hash).toBe("root-1");
    expect(getTreeState(db, 5).root_hash).toBe("root-5");
    expect(getTreeState(db, 10).root_hash).toBe("root-10");
  });

  test("checkpoint_signed_note stores full signed note content", () => {
    const signedNote = `transparency.example.com/log
100
abc123def456789
timestamp 1633024800

— transparency.example.com wXyz123abc...`;

    recordTreeState(db, {
      tree_size: 100,
      root_hash: "abc123def456789",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: signedNote,
    });

    const state = getTreeState(db, 100);
    expect(state.checkpoint_signed_note).toBe(signedNote);
    expect(state.checkpoint_signed_note).toContain("transparency.example.com/log");
    expect(state.checkpoint_signed_note).toContain("— transparency.example.com");
  });

  test("empty history returns empty array", () => {
    const history = getTreeStateHistory(db);
    expect(history).toEqual([]);
  });

  test("concurrent tree size updates maintain consistency", () => {
    // Simulate multiple registrations
    for (let i = 1; i <= 100; i++) {
      updateTreeSize(db, i);
    }

    expect(getCurrentTreeSize(db)).toBe(100);
  });

  test("tree state updated_at is automatically set", () => {
    recordTreeState(db, {
      tree_size: 10,
      root_hash: "hash-10",
      checkpoint_storage_key: "checkpoint",
      checkpoint_signed_note: "note-10",
    });

    const state = getTreeState(db, 10);
    expect(state.updated_at).toBeTruthy();
    expect(typeof state.updated_at).toBe("string");
  });
});
